lib/ruby_indexer/lib/ruby_indexer/uri.rb

```diff
@@ -0,0 +1,108 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyIndexer
+  # The indexer Uri class is a specialization of the regular URI class. It includes convenience methods, special
+  # handling to support Windows paths and simplifications for URI elements that are not used in an LSP context.
+  #
+  # For example, it does not do anything with the `host` portion of the URI, but we include a `require_path` attribute,
+  # so that we know how this URI is mapped in the `$LOAD_PATH` if applicable
+  class Uri < URI::Generic
+    class << self
+      extend T::Sig
+
+      sig { params(path: String, load_path_entry: T.nilable(String)).returns(T.nilable(String)) }
+      def require_from_load_path(path, load_path_entry)
+        return unless load_path_entry
+
+        require_path = path.delete_prefix("#{load_path_entry}/")
+        require_path.delete_suffix!(".rb")
+        require_path
+      end
+
+      sig { params(path: String, load_path_entry: T.nilable(String)).returns(FileUri) }
+      def file(path, load_path_entry = nil)
+        FileUri.new(path: path, require_path: require_from_load_path(path, load_path_entry))
+      end
+    end
+
+    extend T::Sig
+
+    sig { returns(T.nilable(String)) }
+    attr_reader :require_path
+
+    sig do
+      params(
+        scheme: T.nilable(String),
+        path: T.nilable(String),
+        opaque: T.nilable(String),
+        fragment: T.nilable(String),
+        require_path: T.nilable(String),
+      ).void
+    end
+    def initialize(scheme: "file", path: nil, opaque: nil, fragment: nil, require_path: nil)
+      # On Windows, if the path begins with the disk name, we need to add a leading slash to make it a valid URI
+      escaped_path = if !path
+        nil
+      elsif /^[A-Z]:/i.match?(path)
+        URI::DEFAULT_PARSER.escape("/#{path}")
+      elsif path.start_with?("//?/")
+        # Some paths on Windows start with "//?/". This is a special prefix that allows for long file paths
+        URI::DEFAULT_PARSER.escape(path.delete_prefix("//?"))
+      else
+        URI::DEFAULT_PARSER.escape(path)
+      end
+
+      # scheme, userinfo, host, port, registry, path, opaque, query, fragment, parser, arg_check
+      super(scheme, nil, nil, nil, nil, escaped_path, opaque, nil, fragment, URI::DEFAULT_PARSER, true)
+      @require_path = require_path
```

Comment:
I can't find any `Uri` being initialized with `require_path` outside of tests. It looks like only `FileUri` could have this attribute present? If that's the case, should this be `FileUri`'s additional attribute instead?